components/linking.ts

```diff
@@ -0,0 +1,473 @@
+import {NavigationState, PartialState} from '@react-navigation/native';
+import {TabNavigatorParamList} from 'routes';
+import {AvalancheCenterID, AvalancheCenterWebsites, reverseLookup} from 'types/nationalAvalancheCenter';
+
+export const getStateFromUrl = (rawUrl: string): undefined | PartialState<NavigationState<TabNavigatorParamList>> => {
+  const url = new URL(rawUrl);
+  const center: string | undefined = reverseLookup(AvalancheCenterWebsites, url.protocol + '//' + url.hostname + '/');
+  if (!center) {
+    throw new Error(`Unknown hostname for URL: ${rawUrl}`);
+  }
+  const avalancheCenter = center as AvalancheCenterID;
+  const prefixes = prefixesForCenter(avalancheCenter);
+
+  if (prefixes.avalanche) {
+    const forecast = forecastState(avalancheCenter, url.pathname, prefixes.avalanche);
+    if (forecast) {
+      return forecast;
+    }
+  }
+
+  if (prefixes.weatherStation) {
+    const weatherTab = weatherTabState(avalancheCenter, url.pathname, prefixes.weatherStation);
+    if (weatherTab) {
+      return weatherTab;
+    }
+
+    const weatherStation = weatherStationState(avalancheCenter, url.pathname, prefixes.weatherStation);
+    if (weatherStation) {
+      return weatherStation;
+    }
+  }
+
+  if (prefixes.nwacWeatherStation) {
+    const weatherTab = nwacWeatherTabState(avalancheCenter, url.pathname, prefixes.nwacWeatherStation);
+    if (weatherTab) {
+      return weatherTab;
+    }
+
+    const weatherStation = nwacWeatherStationState(avalancheCenter, url.pathname, prefixes.nwacWeatherStation);
+    if (weatherStation) {
+      return weatherStation;
+    }
+  }
+
+  if (prefixes.observations) {
+    // https://nwac.us/observations
+    // https://nwac.us/observations/#/view/observations
+    const observationsTab = observationsTabState(avalancheCenter, url.pathname, prefixes.observations);
+    if (observationsTab) {
+      return observationsTab;
+    }
+
+    // https://nwac.us/observations/#/view/observations/e97eb763-61a6-4188-abc0-a115f5660fe4
+    // https://nwac.us/observations/#/observation/e97eb763-61a6-4188-abc0-a115f5660fe4
+    const observation = observationState(avalancheCenter, url.pathname, prefixes.observations);
+    if (observation) {
+      return observation;
+    }
+  }
+
+  return undefined;
+};
+
+type prefixes = {
+  avalanche?: string;
+  weatherStation?: string;
+  nwacWeatherStation?: string;
+  observations?: string;
+};
+
+// centers use the widgets, and the widgets manage state after some prefix in the URL; not ever center has the same prefix for every widget
+const prefixesForCenter = (avalancheCenter: AvalancheCenterID): prefixes => {
```

Comment:
I think it would make sense for this to live in a separate file vs with all this linking